File: 백준/Gold/3217. malloc/malloc.java
import java.io.*;
import java.util.*;

public class Main {
    static int N;
    static class Node {
        String var;
        int start, end;
        Node prev, next;
        Node(String var, int start, int end) {
            this.var = var;
            this.start = start;
            this.end = end;
        }
    }
    static Node start = new Node("start", 0, 0);
    static Node end = new Node("end", 100001, 100001);
    static HashMap<String, Node> map = new HashMap<>();
    static void init() {
        map.put("start", start);
        map.put("end", end);
        start.next = end;
        end.prev = start;
    }
    static void malloc(String var, int size) {
        for(Node node = start; node != end; node = node.next) {
            int s = node.end + 1;
            int e = node.next.start - 1;
            int length = e - s + 1;
            if(size <= length) {
                Node newNode = new Node(var, s, s + size - 1);
                Node next = node.next;
                node.next = newNode;
                newNode.prev = node;
                next.prev = newNode;
                newNode.next = next;
                map.put(var, newNode);
                return;
            }
        }
        map.put(var, new Node(var, 0, 0));
        return;
    }
    static void free(String var) {
        Node node = map.getOrDefault(var, start);
        if(node.prev == null || node.next == null) return;
        Node prev = node.prev;
        Node next = node.next;
        prev.next = next;
        node.prev = null;
        next.prev = prev;
        node.next = null;
        node.start = 0;
        node.end = 0;
    }
    static int print(String var) {
        Node node = map.getOrDefault(var, start);
        return node.start;
    }
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        // StringTokenizer st = null;
        init();
        N = Integer.parseInt(br.readLine());
        for(int i = 0; i < N; i++) {
            String query = br.readLine();
            if(query.contains("malloc")) {
                String var = query.substring(0,4);
                int size = Integer.parseInt(query.substring(12, query.length() - 2));
                malloc(var, size);
            } else if(query.contains("free")) {
                String var = query.substring(5, query.length() - 2);
                free(var);
            } else {
                String var = query.substring(6, query.length() - 2);
                sb.append(print(var)).append('\n');
            }
        }
        System.out.print(sb);
        br.close();
    }
}

File: 백준/Gold/3217. malloc/malloc.cc
#include <iostream>
#include <unordered_map>
#include <string>

using namespace std;

struct Node {
  string var;
  int start, end;
  Node *prev, *next;
  Node(string var, int start, int end) : var(var), start(start), end(end) {prev = nullptr; next = nullptr;}
};

int N;
Node *start_node = new Node("start", 0, 0);
Node *end_node = new Node("end", 100001, 100001);
unordered_map<string, Node*> map;

void init() {
  map.insert({"start", start_node});
  map.insert({"end", end_node});
  start_node->next = end_node;
  end_node->prev = start_node;
}

void malloc(string var, int size) {
  for(Node *node = start_node; node != end_node; node = node->next) {
    int s = node->end + 1;
    int e = node->next->start - 1;
    int length = e - s + 1;
    if(size <= length) {
      Node *new_node = new Node(var, s, s + size - 1);
      Node *next = node->next;
      node->next = new_node;
      new_node->prev = node;
      next->prev = new_node;
      new_node->next = next;
      map[var] = new_node;
      return;
    }
  }
  map[var] = start_node;
  return;
}

void free(string var) {
  Node *node = map.count(var) ? map[var] : start_node;
  if(node->prev && node->next) {
    Node *prev = node->prev;
    Node *next = node->next;
    prev->next = next;
    next->prev = prev;
    node->prev = nullptr;
    node->prev = nullptr;
    node->start = 0;
    node->end = 0;
  }
}

void print(string var) {
  Node *node = map.count(var) ? map[var] : start_node;
  cout << node->start << '\n';
}

int main() {
  cin.tie(0)->sync_with_stdio(0);
  cin >> N;
  init();
  string query;
  for(int i = 0; i < N; i++) {
    cin >> query;
    if (query.find("malloc") != string::npos) {
      string var = query.substr(0, 4);
      int size = stoi(query.substr(12, query.length() - 13));
      malloc(var, size);
    } else if (query.find("print") != string::npos) {
      string var = query.substr(6, 4);
      print(var);
    } else {
      string var = query.substr(5, 4);
      free(var);
    }
  }
  return 0;
}